react/components/Slider.js

```diff
@@ -1,7 +1,393 @@
+    }
+  }
+
+  updateAfterDrag = () => {
+    const { threshold, currentSlide } = this.props
+    const movement = this.drag.endX - this.drag.startX
+    const movementDistance = Math.abs(movement)
+    const howManySliderToSlide = Math.ceil(movementDistance / (this.selectorWidth / this.perPage))
+
+    const slideToNegativeClone = movement > 0 && currentSlide - howManySliderToSlide < 0
+    const slideToPositiveClone = movement < 0 && currentSlide + howManySliderToSlide > this.innerElements.length - this.perPage
+
+    let newCurrentSlide = currentSlide
+    if (movement > 0 && movementDistance > threshold && this.innerElements.length > this.perPage) {
+      newCurrentSlide = this.prev(howManySliderToSlide)
+    } else if (movement < 0 && movementDistance > threshold && this.innerElements.length > this.perPage) {
+      newCurrentSlide = this.next(howManySliderToSlide)
+    }
+    this.slideToCurrent(slideToNegativeClone || slideToPositiveClone, newCurrentSlide)
+  }
+
+  clearDrag = () => {
+    this.drag = {
+      startX: 0,
+      endX: 0,
+      startY: 0,
+      letItGo: null
+    }
+  }
+
+  setStyle(target, styles) {
+    Object.keys(styles).forEach(attr => {
+      target.style[attr] = styles[attr]
+    })
+  }
+
+  // TODO add touch events
+
+  onMouseDown = e => {
+    e.preventDefault()
+    e.stopPropagation()
+    this.pointerDown = true
+    this.drag.startX = e.pageX
+  }
+
+  onMouseUp = e => {
+    e.stopPropagation()
+    this.pointerDown = false
+    this.enableTransition({ cursor: '-webkit-grab' })
+
+    if (this.drag.endX) {
+      this.updateAfterDrag()
+    }
+
+    this.clearDrag()
+  }
+
+  onMouseMove = e => {
+    const { easing, loop, currentSlide } = this.props
+
+    e.preventDefault()
+    if (this.pointerDown) {
+      // TODO prevent link clicks
+
+      this.drag.endX = e.pageX
+
+      const computedCurrentSlide = loop ? currentSlide + this.perPage : currentSlide
+      const currentOffset = computedCurrentSlide * (this.selectorWidth / this.perPage)
+      const dragOffset = (this.drag.endX - this.drag.startX)
+      const offset = currentOffset - dragOffset
+
+      this.setStyle(this.sliderFrame, {
+        cursor: '-webkit-grabbing',
+        webkitTransition: `all 0ms ${easing}`,
+        transition: `all 0ms ${easing}`,
+        [this.transformProperty]: `translate3d(${offset * -1}px, 0, 0)`
+      })
+    }
+  }
+
+  onMouseLeave = e => {
+
+    if (this.pointerDown) {
+      this.pointerDown = false
+      this.drag.endX = e.pageX
+
+      this.enableTransition({ cursor: '-webkit-grab' })
+      this.updateAfterDrag()
+      this.clearDrag()
+    }
+  }
+
+  render() {
+    const { children: childrenProp, loop } = this.props
+    if (!this.perPage) {
+      this.perPage = resolveSlidesNumber(this.props.perPage)
+    }
+
+    const newChildren = loop ? Children.map([
+      ...childrenProp.slice(childrenProp.length - this.perPage, childrenProp.length),
```

Comment:
shouldn't you call `Children.toArray` before trying to manipulate the prop as one?

Reply:
I'm not sure if this code will stay here, because I might not need it on render since what is does is duplicate the the slides, so I decided to keep it that way until I find a good place for it. And this have a bug too on loop, there is a specific configuration in loop mode that shows the end of the slides.

Reply:
ah, ok then